### PR TITLE
fix(nsis): migrate old shortcuts by setting the path only instead of re-creating a new one

### PR DESCRIPTION
**Pull request by Legend-Master**
References: #9375, #9909

### CODE REVIEW

File: .changes/nsis-migrate-shortcut.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+"tauri-bundler": "patch:bug"
```
Comment (amrbashir) [inline .changes/nsis-migrate-shortcut.md]:
```suggestion
"tauri-bundler": "patch:enhance"
```

File: tooling/bundler/src/bundle/windows/templates/installer.nsi [hunk 2]
```diff
@@ -347,7 +347,7 @@ Var AppStartMenuFolder
 !define MUI_FINISHPAGE_SHOWREADME_FUNCTION CreateDesktopShortcut
 ; Show run app after installation.
 !define MUI_FINISHPAGE_RUN "$INSTDIR\${MAINBINARYNAME}.exe"
-!define MUI_PAGE_CUSTOMFUNCTION_PRE SkipIfPassiveButUpdateShortcutIfUpdate
+!define MUI_PAGE_CUSTOMFUNCTION_PRE SkipIfPassive
```
Comment (amrbashir) [inline tooling/bundler/src/bundle/windows/templates/installer.nsi:350]: why was this changed?

File: tooling/bundler/src/bundle/windows/templates/installer.nsi [hunk 3]
```diff
@@ -577,6 +577,10 @@ Section Install
   WriteRegDWORD SHCTX "${UNINSTKEY}" "NoRepair" "1"
   WriteRegDWORD SHCTX "${UNINSTKEY}" "EstimatedSize" "${ESTIMATEDSIZE}"
 
+  ; We used to use product name as MAINBINARYNAME
+  ; migrate old shortcuts to target the new MAINBINARYNAME
+  Call MigrateProductNameExeShortcuts
```
Comment (amrbashir) [inline tooling/bundler/src/bundle/windows/templates/installer.nsi]: we don't need to create a separate function for migration, in `CreateStartMenuShortcut` and `CreateDesktopShortcut` just check if it exists and update it or re-create it

**Review by amrbashir — CHANGES_REQUESTED** [review]

File: tooling/bundler/src/bundle/windows/templates/installer.nsi (hunk 2 above)
Comment (Legend-Master) [reply tooling/bundler/src/bundle/windows/templates/installer.nsi:350]: wasn't this change for migrating shortcuts? since we do it on install now, no need to do it here anymore

File: tooling/bundler/src/bundle/windows/templates/installer.nsi (hunk 3 above)
Comment (Legend-Master) [reply tooling/bundler/src/bundle/windows/templates/installer.nsi]: it doesn't make sense to be in create shortcuts functions, we might skip those functions (we might not currently, but from the name it's not clear it's also responsible for migrating old shortcuts)

File: tooling/bundler/src/bundle/windows/templates/installer.nsi (hunk 2 above)
Comment (amrbashir) [reply tooling/bundler/src/bundle/windows/templates/installer.nsi:350]: No, this is for passive mode installer where it is not necessarily an update

Comment (amrbashir) [reply tooling/bundler/src/bundle/windows/templates/installer.nsi:350]: Actually I realize my implementation and name was wrong

File: tooling/bundler/src/bundle/windows/templates/installer.nsi (hunk 3 above)
Comment (amrbashir) [reply tooling/bundler/src/bundle/windows/templates/installer.nsi]: The name could change, and instead of multiple functions we can just put all logic related to a shortcut inside one function, so we always call it but the logic inside it decides whether to create, update or skip

File: tooling/bundler/src/bundle/windows/templates/installer.nsi (hunk 2 above)
Comment (amrbashir) [reply tooling/bundler/src/bundle/windows/templates/installer.nsi:350]: Let's ignore this and remove, my brain can't handle more NSIS this week

Comment (Legend-Master) [reply tooling/bundler/src/bundle/windows/templates/installer.nsi:350]: isn't passive mode installer handled in install, line 592?

File: tooling/bundler/src/bundle/windows/templates/installer.nsi (hunk 3 above)
Comment (Legend-Master) [reply tooling/bundler/src/bundle/windows/templates/installer.nsi]: kk

Comment (Legend-Master) [reply tooling/bundler/src/bundle/windows/templates/installer.nsi]: Wait, we still need to separate start menu shortcut from desktop one, they have different configs from UIs

File: tooling/bundler/src/bundle/windows/templates/installer.nsi (hunk 2 above)
Comment (amrbashir) [reply tooling/bundler/src/bundle/windows/templates/installer.nsi:350]: yeah I think so, but I remember I had a case where ot wasn't covered by that and I added that code, I may have changed the code a bit after that. I really don't remember, let's just move on

File: tooling/bundler/src/bundle/windows/templates/installer.nsi (hunk 3 above)
Comment (amrbashir) [reply tooling/bundler/src/bundle/windows/templates/installer.nsi]: Yeah 2 functions, one for desktop and one for start menu. I think we need to extract some of the logic into another macro

Comment (Legend-Master) [reply tooling/bundler/src/bundle/windows/templates/installer.nsi]: To be honest, I still don't think it make much sense to put migration logic in `CreateStartMenuShortcut`/`CreateDesktopShortcut`, since this is not about creating the shortcut at all (changing those function's name to something like `HandleStartMenuShortcut` doesn't make much sense to me either), migration should just be a separate function (if you don't like a function we can use a macro but I don't think that's what you mean)

Comment (amrbashir) [reply tooling/bundler/src/bundle/windows/templates/installer.nsi]: Problem is to call said functions, we will do the conditional at the call site which sure is okay but bloats the installation section and I would love to hide these in a function so it is easier to focus on just that function and understand the logic without need of outer context that much.

Comment (Legend-Master) [reply tooling/bundler/src/bundle/windows/templates/installer.nsi]: Do you mean extract this to a function?

https://github.com/tauri-apps/tauri/blob/040289ea52467c423bd895026d891553409e862a/tooling/bundler/src/bundle/windows/templates/installer.nsi#L582-L595

Comment (amrbashir) [reply tooling/bundler/src/bundle/windows/templates/installer.nsi]: I think I would like it to be
```nsi 
   ; Create start menu shortcut 
   !insertmacro MUI_STARTMENU_WRITE_BEGIN Application 
     Call CreateOrUpdateStartMenuShortcut 
   !insertmacro MUI_STARTMENU_WRITE_END 
  
   ; Create desktop shortcut
   Call CreateOrUpdateDesktopShortcut
```

and each of these functions would hide the implementation inside it

Comment (Legend-Master) [reply tooling/bundler/src/bundle/windows/templates/installer.nsi]: This wouldn't work, start menu one will be skipped on update

Migration logic here is very different than creating the shortcuts in this context, I suggest we keep shortcut migration code at a central place

Comment (amrbashir) [reply tooling/bundler/src/bundle/windows/templates/installer.nsi]: > This wouldn't work, start menu one will be skipped on update

Why would it be skipped?

Comment (Legend-Master) [reply tooling/bundler/src/bundle/windows/templates/installer.nsi]: Doesn't this skip the inner block if no start menu page present (e.g. on update)?

```nsis
   !insertmacro MUI_STARTMENU_WRITE_BEGIN Application 
     Call CreateOrUpdateStartMenuShortcut 
   !insertmacro MUI_STARTMENU_WRITE_END 
```

Comment (amrbashir) [reply tooling/bundler/src/bundle/windows/templates/installer.nsi]: it still calls it, we just skipped page, so think of it as the page was shown for a split second and then clicked next immediately, so the default values are used.

Comment (Legend-Master) [reply tooling/bundler/src/bundle/windows/templates/installer.nsi]: You're right, I thought skip if passive means not to run that part, it's actually more or less the same as clicking the next button

But still, it will not be the case anymore once we add in an option to disable that page, I still think put the migration logic together make a lot more sense than putting them in create functions (yes, the name can change, but the caller needs to be more cautious about skipping those function)

Comment (amrbashir) [reply tooling/bundler/src/bundle/windows/templates/installer.nsi]: That's what I am trying to avoid, by always calling them and let the function itself handle the skipping.

Comment (Legend-Master) [reply tooling/bundler/src/bundle/windows/templates/installer.nsi]:
```nsis
  ; Create start menu shortcut
  !insertmacro MUI_STARTMENU_WRITE_BEGIN Application
    Call CreateStartMenuShortcut
  !insertmacro MUI_STARTMENU_WRITE_END

  ; Create desktop shortcut for silent and passive installers
  ; because finish page will be skipped
  IfSilent create_shortcut 0
  StrCmp $PassiveMode "1" create_shortcut shortcut_done
  create_shortcut:
    Call CreateDesktopShortcut
  shortcut_done:
```

This part itself does some checking and might skip it (it might not, but let's just get it done, in the most clear way, I don't want to play with nsis anymore)

NSIS itself is so error prone (that's why we use rust to write most things not nsi script), I really don't want to rely on those things to work, calling migration here is most clear way of always running it

Comment (amrbashir) [reply tooling/bundler/src/bundle/windows/templates/installer.nsi]: That's the thing we don't always need to migrate, like normal install without update should always recreate not update.

Comment (Legend-Master) [reply tooling/bundler/src/bundle/windows/templates/installer.nsi]: We should always migrate, even the user checks no desktop shortcut at the end, we should still update existing shortcuts

Comment (amrbashir) [reply tooling/bundler/src/bundle/windows/templates/installer.nsi]: I don't agree with that, we shouldn't update if we are not in update mode, otherwise just recreate, reason being if somehow the user messed up their shortcuts and wasn't very experienced to fix it, they will usually resort to reinstalling which should recreate the shortcut and fix it. They could go through the trouble of uninstalling then installing but that could potentially be more destructive.

Comment (Legend-Master) [reply tooling/bundler/src/bundle/windows/templates/installer.nsi]: Current implementation will do exactly that, if user re-runs the installer, it will first migrate the shortcut, and then re-create it, if the user check no update mode, it will only do the migration

Comment (amrbashir) [reply tooling/bundler/src/bundle/windows/templates/installer.nsi]: I know and that is redundant, we could be avoiding that extra work.

Comment (Legend-Master) [reply tooling/bundler/src/bundle/windows/templates/installer.nsi]: Alright, let's don't get stuck here debating, I still think it's better to put migration code in a function, but I'll change it to doing it in create shortcut function

File: tooling/bundler/src/bundle/windows/templates/installer.nsi [hunk 4]
```diff
@@ -727,43 +736,39 @@ Function SkipIfPassive
...
 FunctionEnd
 
-Function SkipIfPassiveButUpdateShortcutIfUpdate
-  ${If} $PassiveMode = 1
-    Call CreateDesktopShortcut
-    Abort
+Function CreateOrUpdateStartMenuShortcut
+  ; We used to use product name as MAINBINARYNAME
+  ; migrate old shortcuts to target the new MAINBINARYNAME
+  ${If} ${FileExists} "$DESKTOP\${PRODUCTNAME}.lnk"
```
Comment (amrbashir) [inline tooling/bundler/src/bundle/windows/templates/installer.nsi:742]: We should only update if in update mode, otherwise recreate the shortcut

```suggestion
  ; Update shortcuts target path if updating
  ${If} $UpdateMode = 1
  ${AndIf} ${FileExists} "$DESKTOP\${PRODUCTNAME}.lnk"
```

File: tooling/bundler/src/bundle/windows/templates/installer.nsi [hunk 5]
```diff
@@ -727,43 +736,39 @@ Function SkipIfPassive
...
-Function CreateStartMenuShortcut
-  ; Skip creating shortcut if in update mode.
-  ; See `CreateDesktopShortcut` above.
-  ${If} $UpdateMode = 1
-    IfFileExists "$SMPROGRAMS\$AppStartMenuFolder\${PRODUCTNAME}.lnk" +2 0
-      Return
+Function CreateOrUpdateDesktopShortcut
+  ; We used to use product name as MAINBINARYNAME
+  ; migrate old shortcuts to target the new MAINBINARYNAME
+  ${If} ${FileExists} "$SMPROGRAMS\$AppStartMenuFolder\${PRODUCTNAME}.lnk"
```
Comment (amrbashir) [inline tooling/bundler/src/bundle/windows/templates/installer.nsi:761]: ditto
```suggestion
  ; Update shortcuts target path if updating
  ${If} $UpdateMode = 1
  ${AndIf} ${FileExists} "$SMPROGRAMS\$AppStartMenuFolder\${PRODUCTNAME}.lnk"
```

**Review by amrbashir — CHANGES_REQUESTED** [review]

File: tooling/bundler/src/bundle/windows/templates/installer.nsi (hunk 4 above)
Comment (Legend-Master) [reply tooling/bundler/src/bundle/windows/templates/installer.nsi:742]: We don't pass in /update in the earlier versions, and this migration is exactly for those versions, we shouldn't re-create if /update is not present in the command line args

Comment (amrbashir) [reply tooling/bundler/src/bundle/windows/templates/installer.nsi:742]: right, I forgot if we decided on a solution for this tbh, let's roll with this solution for now as it seems the safest

**Review by amrbashir — APPROVED** [review]
